### ISSUE TITLE
Out of order `$n` params get incorrect values

### ISSUE THREAD

**Issue by snej** [mattn/go-sqlite3#1169]
I have code that works with the `modernc.org/sqlite` binding and wanted to move it over to this one. I figured I could just change the `import` line and most everything would work ... but in fact many of my tests broke badly. After some debugging, I finally figured out that **this binding doesn't apply numbered SQL query parameters (like `$2`) correctly if they don't appear in order in the query.** Instead the first `$` parameter is always given the first Go argument value, and so on, regardless of its number.

(Out-of-order parameters can happen if the same `Query` or `Exec` call is used with several different SQL strings that use its parameters in a different order -- I've got several places where I do this. It can also happen if you alter a query in a way that moves a clause containing parameters from one place to another. Also, if a query uses a parameter more than once, then obviously one of those occurrences won't match its position in the Go argument list.)

### Example:

> `db.Exec("update foo set id = $2 where id=$1", 456, 789)`

**Expected:** Query evaluates as `update foo set id = 789 where id=456`.

**Actually:** Query evaluates as `update foo set id = 456 where id=789`.

### Test Case

```go

func TestOutOfOrderParams(t *testing.T) {
	db, err := sql.Open("sqlite3", "file:/?mode=memory")
	if err != nil {
		t.Fatal("Failed to open db:", err)
	}

	_, err = db.Exec("create table foo (id integer)")
	if err != nil {
		t.Fatal("Failed to create table:", err)
	}

	_, err = db.Exec("insert into foo(id) values(123)")
	if err != nil {
		t.Fatal("Failed to insert record:", err)
	}

	// Update the row.
	res, err := db.Exec("update foo set id = $1 where id=$2", 456, 123)
	if err != nil {
		t.Fatal("Failed to update record:", err)
	}
	affected, _ := res.RowsAffected()
	if err != nil {
		t.Fatal("Failed to get RowsAffected:", err)
	}
	if affected != 1 {
		t.Fatalf("Expected %d for affected rows, but %d:", 1, affected)
	}

	// Update again, but this time the SQL params are out of order.
	res, err = db.Exec("update foo set id = $2 where id=$1", 456, 789)
	if err != nil {
		t.Fatal("Failed to update record:", err)
	}
	affected, _ = res.RowsAffected()
	if err != nil {
		t.Fatal("Failed to get RowsAffected:", err)
	}
	if affected != 1 {
		t.Fatalf("Expected %d for affected rows, but %d:", 1, affected).   // <-- test fails here
	}
}
```

**Comment by snej** [comment]
I went back to the [SQLite docs](https://sqlite.org/c3ref/bind_blob.html) and noticed that the only form for bindings that takes a numeric index is `?`:

>literals may be replaced by a parameter that matches one of following templates:
>`?`
>`?NNN`
>`:VVV`
>`@VVV`
>`$VVV`
>In the templates above, NNN represents an integer literal, and VVV represents an alphanumeric identifier.

Once I changed my `$1` etc parameters to `?1` etc, the problem went away.

So at some level this is my fault, but it's interesting that this worked fine with the modernc driver. It was somehow able to get the args bound correctly. I don't know whether you want to look into that improvement...
